### PR TITLE
[WIP] Support external services

### PR DESCRIPTION
**Pull request by ijsnellf**
Proof of concept for external services.

Needs:
- [x] Formatting + refactoring
- [x] Tests for in/out port
- [ ] Static discovery
- [ ] DNS discovery

### CODE REVIEW

File: pilot/proxy/envoy/config.go [hunk 1]
```diff
@@ -934,13 +939,98 @@ func buildEgressHTTPRoutes(mesh *meshconfig.MeshConfig, node proxy.Node,
...
+	if node.Type == proxy.Router {
+		return listeners, clusters
+	}
+
+	// TODO: port conflicts?
+	for _, fsConfig := range config.ForeignServices() {
+		fs := fsConfig.Spec.(*routingv2.ForeignService)
+		for _, port := range fs.Ports {
+			protocol := model.ConvertCaseInsensitiveStringToProtocol(port.Protocol)
+			if !model.IsEgressRulesSupportedTCPProtocol(protocol) {
```
Comment (rshriram) [inline pilot/proxy/envoy/config.go]: why? We can still do plain tcp proxy right? In fact, thats all we would do, except for mongo

Comment (ijsnellf) [reply pilot/proxy/envoy/config.go]: this filters out http, http2 and grpc ports. yes, we only support tcp and mongo here.

File: pilot/proxy/envoy/testdata/foreign-service-rule-tcp.yaml.golden [hunk 2]
```diff
@@ -0,0 +1,7 @@
+hosts:
+  - "172.217.0.0/16"
```
Comment (mandarjog) [inline pilot/proxy/envoy/testdata/foreign-service-rule-tcp.yaml.golden]: updated name?

Comment (ijsnellf) [reply pilot/proxy/envoy/testdata/foreign-service-rule-tcp.yaml.golden]: thanks, missed those ones

**Comment by ijsnellf** [comment]
Closing this PR in favor of a new PR for external services supporting all discovery types and removing default TLS origination for HTTPS.

**Pull request closed without merging**